feat(analysis/normed_space) bundled bounded linear maps

jlpaca opened the pull request:
- `bounded_linear_map` as a structure.
- some elementary properties of the operator norm.

TO CONTRIBUTORS:

Make sure you have:

  * [ ] reviewed and applied the coding style: [coding](https://github.com/leanprover/mathlib/blob/master/docs/style.md), [naming](https://github.com/leanprover/mathlib/blob/master/docs/naming.md)
  * [ ] for tactics:
     * [ ] added or adapted documentation in [tactics.md](https://github.com/leanprover/mathlib/blob/master/docs/tactics.md)
     * [ ] write an example of use of the new feature in [tactics.lean](https://github.com/leanprover/mathlib/blob/master/tests/tactics.lean)
  * [ ] make sure definitions and lemmas are put in the right files
  * [ ] make sure definitions and lemmas are not redundant

If this PR is related to a discussion on Zulip, please include a link in the discussion.

For reviewers: [code review check list](https://github.com/leanprover/mathlib/blob/master/docs/code-review.md)

jlpaca committed:
feat(src/analysis/normed_space): bundled bounded linear maps

PatrickMassot commented on src/analysis/normed_space/bounded_linear_maps.lean:
```diff
@@ -22,15 +22,115 @@ local notation f ` →_{`:50 a `} `:0 b := filter.tendsto f (nhds a) (nhds b)
+  (hf _) (mul_le_mul_of_nonneg_right (le_max_right _ _) (norm_nonneg _))⟩
+
+-- some lemmas about things having bounds and things being bounds
+lemma ratio_has_pos_bound : ∃ M > 0, ∀ x ≠ 0, ∥f x∥ / ∥x∥ ≤ M :=
```
Isn't `x ≠ 0` a useless assumption?

jlpaca replied:
fixed in 8946ddcaee6ff2c15022befd6b1668cb11c2fca5 . thanks!

PatrickMassot commented on src/analysis/normed_space/bounded_linear_maps.lean:
```diff
@@ -22,15 +22,115 @@ local notation f ` →_{`:50 a `} `:0 b := filter.tendsto f (nhds a) (nhds b)
+
+instance : has_neg (bounded_linear_map k E F) := ⟨λ f, (-1 : k) • f⟩
+instance : has_sub (bounded_linear_map k E F) := ⟨λ f g, f + (-g)⟩
+
```
Can't we state a vector space instance then? Maybe even an algebra when E=F.

jlpaca replied:
both of those instances exist (without breaking things elsewhere) as of 9a3cce0983597bb1cccc2e6fdbb7940ddccaa22a.

PatrickMassot commented on src/analysis/normed_space/operator_norm.lean:
```diff
@@ -16,9 +16,89 @@ particular
+    (λ c ⟨_, hc⟩, div_le_of_le_mul hlt (by rw mul_comm; exact hc _))))
+
+-- results about bounding the unit ball. (naming conventions?)
+lemma ratio_le_op_norm: ∥x∥ > 0 → ∥f x∥ / ∥x∥ ≤ ∥f∥ :=
```
Again, the assumption is useless, right?

jlpaca replied:
fixed in 8946ddcaee6ff2c15022befd6b1668cb11c2fca5.

PatrickMassot commented on src/analysis/normed_space/operator_norm.lean:
```diff
@@ -16,9 +16,89 @@ particular
+            (by rw div_mul_eq_mul_div; exact le_div_of_mul_le hlt
+            (by rw [ mul_comm, ←norm_smul ]; exact hc _))⟩))
+        (λ heq, by rw [←heq, zero_mul]; exact hn))))
+
```
Why don't we get a normed space of bounded linear maps?

jlpaca replied:
`bounded_linear_map` are now a normed space in a80dd8de3f8b19cfb68b789b95d044ade85fbe48.

PatrickMassot reviewed (commented):
I left some minor comments

jlpaca committed:
fix(analysis/normed_space): remove useless assumptions

jlpaca committed:
feat(analysis/normed_space): add instances

- bounded_linear_map is a k-vector space
- bounded_linear_map together with op_norm are a normed space

jlpaca committed:
feat(analysis/normed_space/bounded_linear_maps): add instance for the endomorphism algebra

jlpaca committed:
fix(analysis/normed_space): fix type class resolution issue

jlpaca committed:
feat(analysis/normed_space): bounded linear maps are lipschitz continuous

jlpaca committed:
style(analysis/normed_space): remove trailing whitespace

jlpaca committed:
Merge branch 'master' of https://github.com/leanprover-community/mathlib into operator-norm

jcommelin commented on src/analysis/normed_space/bounded_linear_maps.lean:
```diff
@@ -22,15 +23,162 @@ local notation f ` →_{`:50 a `} `:0 b := filter.tendsto f (nhds a) (nhds b)
+
+variables {α : k} {u v : E}
+
+-- (helpful?) lemmas because invoking the corresponding results
```
Indeed such lemmas are helpful. I think the comment doesn't need to end up in mathlib thought ;-)
```suggestion
```

jcommelin commented on src/analysis/normed_space/bounded_linear_maps.lean:
```diff
@@ -22,15 +23,162 @@ local notation f ` →_{`:50 a `} `:0 b := filter.tendsto f (nhds a) (nhds b)
+variables {α : k} {u v : E}
+
+-- (helpful?) lemmas because invoking the corresponding results
+-- for linear_map isn't straightforward.
```
```suggestion
```

jcommelin commented on src/analysis/normed_space/operator_norm.lean:
```diff
@@ -15,10 +15,102 @@ particular
+variable  {k : Type*}
+variables {E : Type*} {F : Type*} {G : Type*}
+
+-- refactor : formulation for bundled bounded_linear_map
```
What exactly is the intention of this comment? Should it end up in mathlib?

jcommelin commented on src/analysis/normed_space/bounded_linear_maps.lean:
```diff
@@ -22,15 +23,162 @@ local notation f ` →_{`:50 a `} `:0 b := filter.tendsto f (nhds a) (nhds b)
+  ⟨f.to_linear_map.is_linear,
+  let ⟨M, hMg, hMb⟩ := has_pos_bound in ⟨M, hMg, hMb⟩⟩
+
+-- the zero map and the identity maps are bounded
```
If you want to write comments, that is very good. However, it would be even better to write docstrings. Because these docstrings will then show up in tooltips in VScode and such. The docstring has to be relevant for exactly the one statement that follows it. So this comment would have to be split into two docstrings.
You could also do this elsewhere in this PR.

jcommelin commented on src/analysis/normed_space/bounded_linear_maps.lean:
```diff
@@ -22,15 +23,162 @@ local notation f ` →_{`:50 a `} `:0 b := filter.tendsto f (nhds a) (nhds b)
+}
+
+
+-- a bounded linear map is continuous.
```
Docstring

jcommelin reviewed (commented):
Some minor comments about comments

jlpaca committed:
Merge branch 'master' of https://github.com/leanprover/mathlib into operator-norm

- also some notation for bounded_linear_map

jlpaca committed:
doc(analysis/normed_space): docstrings

jcommelin commented on src/analysis/normed_space/bounded_linear_maps.lean:
```diff
@@ -6,9 +6,9 @@ Authors: Patrick Massot, Johannes Hölzl
+import ring_theory.algebra
 import analysis.normed_space.basic
-import analysis.asymptotics
+import ..asymptotics
```
I think this is not conform mathlib style.

jcommelin commented on src/analysis/normed_space/bounded_linear_maps.lean:
```diff
@@ -17,17 +17,168 @@ mul_inv_eq hb ha
+def of_linear_map_of_bounded {f : E →ₗ[k] F}
+  (h : ∃ M, ∀ x, ∥f x∥ ≤ M * ∥x∥) : E →L[k] F := ⟨f, h⟩
+
+/-- the zero map is bounded.     -/
```
```suggestion
/-- The bounded map that is constantly zero. -/
```

jcommelin commented on src/analysis/normed_space/bounded_linear_maps.lean:
```diff
@@ -17,17 +17,168 @@ mul_inv_eq hb ha
+/-- the identity map is bounded. -/
+def id   : E →L[k] E := ⟨linear_map.id, 1, λ x, le_of_eq (one_mul _).symm⟩
+
+/-- the composition of two bounded linear maps is bounded. -/
```
```suggestion
/-- Composition of bounded linear maps. -/
```

jcommelin commented on src/analysis/normed_space/bounded_linear_maps.lean:
```diff
@@ -17,17 +17,168 @@ mul_inv_eq hb ha
+  commutes' := λ _ _, ext $ λ _, map_smul _ _ _,
+}
+
+/-- bounded linear maps are continuous. -/
```
```suggestion
/-- Bounded linear maps are continuous. -/
```

jcommelin commented on src/analysis/normed_space/operator_norm.lean:
```diff
@@ -15,208 +15,99 @@ particular
+
+noncomputable instance : has_norm (bounded_linear_map k E F) := ⟨op_norm⟩
+
+-- so that invocations of real.Inf_le make sense: the set of
```
This comment seems to start halfway a sentence.

jcommelin commented on src/analysis/normed_space/operator_norm.lean:
```diff
@@ -15,208 +15,99 @@ particular
+  (λ hlt, div_le_of_le_mul hlt (by rw mul_comm; exact le_op_norm _))
+  (λ heq, by rw [←heq, div_zero]; exact op_norm_nonneg _))
+
+/-- the image of the unit ball under a bounded linear map is bounded. -/
```
```suggestion
/-- The image of the unit ball under a bounded linear map is bounded. -/
```

jcommelin commented on src/analysis/normed_space/operator_norm.lean:
```diff
@@ -15,208 +15,99 @@ particular
+  (λ heq, by rw [←heq, div_zero]; exact op_norm_nonneg _))
+
+/-- the image of the unit ball under a bounded linear map is bounded. -/
+lemma unit_le_op_norm : ∥x∥ ≤ 1 → ∥f x∥ ≤ ∥f∥ :=
```
I think you might as well move `hx : ||x|| <= 1` to the assumptions.

jcommelin commented on src/analysis/normed_space/operator_norm.lean:
```diff
@@ -15,208 +15,99 @@ particular
+            (by rw [ mul_comm, ←norm_smul ]; exact hc _))⟩))
+        (λ heq, by rw [←heq, zero_mul]; exact hn))))
+
+/-- bounded linear maps themselves form a normed space w/ the op norm -/
```
```suggestion
/-- Bounded linear maps form a normed space with respect to the operator norm -/
```

jcommelin commented on src/analysis/normed_space/operator_norm.lean:
```diff
@@ -15,208 +15,99 @@ particular
+noncomputable instance : normed_space k (bounded_linear_map k E F) :=
+  normed_space.of_core _ _ ⟨op_norm_eq_zero, op_norm_smul, op_norm_triangle⟩
+
+/-- operator norm is submultiplicative. -/
```
```suggestion
/-- The operator norm is submultiplicative. -/
```

jcommelin commented on src/analysis/normed_space/operator_norm.lean:
```diff
@@ -15,208 +15,99 @@ particular
+  λ x, by rw mul_assoc; calc _ ≤ ∥h∥ * ∥f x∥: le_op_norm _
+  ... ≤ _ : mul_le_mul_of_nonneg_left (le_op_norm _) (op_norm_nonneg _)⟩)
+
+/-- bounded linear maps are lipschitz continuous. -/
```
```suggestion
/-- Bounded linear maps are lipschitz continuous. -/
```

jcommelin reviewed (commented):
Some more comments.

jlpaca committed:
refactor(analysis/normed_space): derivatives with bundled bounded linear maps

- chain rule lemma & two results about real normed spaces work in progress

jlpaca committed:
Merge branch 'master' of https://github.com/leanprover/mathlib into operator-norm
